service/gcs/errors/errors.go

```diff
@@ -45,3 +62,111 @@ func (e *processDoesNotExistError) Error() string {
 func NewProcessDoesNotExistError(pid int) *processDoesNotExistError {
 	return &processDoesNotExistError{Pid: pid}
 }
+
+// StackTracer is an interface originating (but not exported) from the
+// github.com/pkg/errors package. It defines something which can return a stack
+// trace.
+type StackTracer interface {
+	StackTrace() errors.StackTrace
+}
+
+type baseHresultError struct {
+	hresult Hresult
+}
+
+func (e *baseHresultError) Error() string {
+	return fmt.Sprintf("HRESULT: %d", e.Hresult())
+}
+func (e *baseHresultError) Hresult() Hresult {
+	return e.hresult
+}
+func (e *baseHresultError) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 'v':
+		fallthrough
+	case 's':
+		io.WriteString(s, e.Error())
+	case 'q':
+		fmt.Fprintf(s, "%q", e.Error())
+	}
+}
+
+type wrappingHresultError struct {
+	cause   error
+	hresult Hresult
+}
+
+func (e *wrappingHresultError) Error() string {
+	return fmt.Sprintf("HRESULT %d", e.Hresult()) + ": " + e.Cause().Error()
+}
+func (e *wrappingHresultError) Hresult() Hresult {
+	return e.hresult
+}
+func (e *wrappingHresultError) Cause() error {
+	return e.cause
+}
+func (e *wrappingHresultError) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 'v':
+		if s.Flag('+') {
+			fmt.Fprintf(s, "%+v\n", e.Cause())
+			return
+		}
+		fallthrough
+	case 's':
+		io.WriteString(s, e.Error())
+	case 'q':
+		fmt.Fprintf(s, "%q", e.Error())
+	}
+}
+func (e *wrappingHresultError) StackTrace() errors.StackTrace {
+	type stackTracer interface {
+		StackTrace() errors.StackTrace
+	}
+	serr, ok := e.Cause().(stackTracer)
```

Comment:
That's fair. I think I'm going to keep it like this to stick to the general pattern I've seen of returning the requested value at the end of a function, and "invalid" values (such as errors or in this case nil) earlier in the function.